blockchain/fullblocktests/generate.go

```diff
@@ -2317,11 +2342,79 @@ func GenerateHeaders() (generator *testGenerator, tests [][]TestInstance) {
 		g.updateBlockState("b3", origHash, "b3", b3)
 	}
 	rejected(blockchain.ErrUnexpectedDifficulty)
-	// Adding a block with valid header
+	// Adding a block with valid header but invalid spend
 	//
-	// ... -> b0() -> b4(1)
+	// ... -> b0() -> b4(2)
 	g.setTip("b0")
-	g.nextBlock("b4", outs[1])
+	g.nextBlock("b4", outs[2])
 	accepted()
+	// Adding a block with valid header and valid spend, but invalid parent
```

Comment:
The comment is wrong here. In headers-first you don't accept a header with an invalid parent. The tip is never reset to `"b0"` and thus it has a valid parent of `"b4"`.

In general, checkpointed headers are assumed to be correct as it's hardcoded into the binary and it's actually a consensus rule (whether or not this is good is besides the point). Once you accept headers, the block **must** match with the header. If it doesn't I believe the node just hangs.

So we can have tests for rejecting headers but once we accept the headers, it's assumed as correct so the blockdata corresponding to the headers should never be rejected.